processor/src/main/java/com/linecorp/decaton/processor/processors/BatchingProcessor.java

```diff
@@ -0,0 +1,141 @@
+
+    private final ScheduledExecutorService executor;
+    private List<BatchingTask<T>> currentBatch = new ArrayList<>();
+    private final long lingerMillis;
+    private final int capacity;
+
+    @Getter
+    @Accessors(fluent = true)
+    public static class BatchingTask<T> {
+        @Getter
+        private final Completion completion;
+        @Getter
+        private final ProcessingContext<T> context;
+        @Getter
+        private final T task;
+
+        private BatchingTask(Completion completion, ProcessingContext<T> context, T task) {
+            this.completion = completion;
+            this.context = context;
+            this.task = task;
+        }
+    }
+
+    /**
+     * Instantiate {@link BatchingProcessor}.
+     * If you only need one limit, please set large enough value to another.
+     * @param lingerMillis time limit for this processor. On every lingerMillis milliseconds,
+     * tasks in past lingerMillis milliseconds are pushed to {@link BatchingTask#processBatchingTasks(List)}.
+     * @param capacity size limit for this processor. Every time tasks’size reaches capacity,
+     * tasks in past before reaching capacity are pushed to {@link BatchingTask#processBatchingTasks(List)}.
+     */
+    protected BatchingProcessor(long lingerMillis, int capacity) {
+        this.lingerMillis = lingerMillis;
+        this.capacity = capacity;
+
+        ScheduledThreadPoolExecutor scheduledExecutor = new ScheduledThreadPoolExecutor(1, r -> {
+            Thread th = new Thread(r);
+            th.setName("Decaton" + BatchingProcessor.class.getSimpleName()
+                       + '/' + System.identityHashCode(this));
+            return th;
+        });
+
+        // For faster shutdown cancel all pending flush on executor shutdown.
+        // In fact for this purpose we have two options here:
+        // A. Cancel all pending flush
+        // B. Run all pending flush immediately, regardless remaining time until flush
+        // A has a downside that it forces consumer to re-process some amount of tasks.
+        // B has a downside that it may causes flush bursting and then on spike for storage/API access in
+        // downstream processor.
+        // Given that this feature is expected to be used mainly for reducing workload for storage/APIs,
+        // we think taking A is much desirable behavior in most cases.
+        scheduledExecutor.setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
+        executor = scheduledExecutor;
+
+        scheduleFlush();
+    }
+
+    private void periodicallyFlushTask() {
+        final List<BatchingTask<T>> batch;
+        synchronized (this) {
+            if (!currentBatch.isEmpty()) {
+                batch = currentBatch;
+                currentBatch = new ArrayList<>();
+            } else {
+                batch = null;
+            }
+        }
+        if (batch != null) {
+            processBatchingTasks(batch);
+        }
+        scheduleFlush();
```

Comment:
H-m, I realized that unlimited number of scheduleFlush task may be queued in executor when `processBatchingTasks` for head-batch is somehow blocked.
Also, I think scheduleFlush task may be queued even when there are no incoming tasks, it sounds wasteful. (then the flush task may return immediately so not gonna be big problem though)

So I guess we need some fixes like:
- Schedule flushTask only when new batch is created
- Don't schedule further flushTask if there's pending one

Reply:
> unlimited number of scheduleFlush task may be queued

This sounds not true because new delayed task will not be queued until this task returns.
Please nevermind my comment.

Reply:
The next `scheduleFlush()` calling is after `processBatchingTasks(batch)` is complete in same thread, so it's ok.
Is this correct? If I have a misunderstood, please point it out.🙇